src/math/ring_signature.cpp

```diff
@@ -203,34 +191,41 @@ ec_secret calculate_s(const ec_secret& k, const ec_secret& e,
     rc = ec_negate(result);
     BITCOIN_ASSERT(rc);
     // result = k - e x
-    ec_add(result, k);
+    rc = ec_add(result, k);
+    BITCOIN_ASSERT(rc);
     return result;
 }
 
-bool sign(
-    ring_signature& out,
-    const secret_list& secrets,
-    const key_rings& rings,
-    const data_slice message,
-    const data_slice seed)
+bool sign(ring_signature& out, const secret_list& secrets,
+    const key_rings& rings, const data_slice message, const data_slice seed)
 {
     // Create public key -> secret key map
     auto secret_keys = generate_keys_map(secrets);
+    if (secret_keys.empty())
+        return false;
     // Organize known public keys into corresponding rings
     const auto known_keys_by_ring =
         partition_keys_into_rings(secret_keys, rings);
+
     // Check we know a secret key in each ring
-    if (!all_rings_nonzero_size(known_keys_by_ring))
+    auto non_empty_ring = [](const point_list& ring)
+    {
+        return !ring.empty();
+    };
+    const bool all_rings_nonzero_size = std::all_of(
+        known_keys_by_ring.begin(), known_keys_by_ring.end(), non_empty_ring);
+
```

Comment:
Suggest avoiding the double negative for readability and instead use “any ring zero size”.